From 07ca5f3b066947e09c7f349a946db5b209731615 Mon Sep 17 00:00:00 2001
From: Joel Chippindale <joel.chippindale@gmail.com>
Date: Fri, 31 Oct 2008 14:37:00 +0000
Subject: [PATCH] Fixed method_missing for ActionMailer

So that it no longer matches methods where deliver or create are not a suffix.

This reverses the change in behaviour made in commit 3cf773b187e803e16b8237e5923fa4c1139cde8a
---
 actionmailer/lib/action_mailer/base.rb |   17 ++++++++++-------
 actionmailer/test/mail_service_test.rb |    4 ++++
 2 files changed, 14 insertions(+), 7 deletions(-)

diff --git a/actionmailer/lib/action_mailer/base.rb b/actionmailer/lib/action_mailer/base.rb
index 043f56b..16161df 100644
--- a/actionmailer/lib/action_mailer/base.rb
+++ b/actionmailer/lib/action_mailer/base.rb
@@ -386,12 +386,15 @@ module ActionMailer #:nodoc:
       end
 
       def method_missing(method_symbol, *parameters) #:nodoc:
-        match = matches_dynamic_method?(method_symbol)
-        case match[1]
-          when 'create'  then new(match[2], *parameters).mail
-          when 'deliver' then new(match[2], *parameters).deliver!
-          when 'new'     then nil
-          else super
+        if match = matches_dynamic_method?(method_symbol)
+          case match[1]
+            when 'create'  then new(match[2], *parameters).mail
+            when 'deliver' then new(match[2], *parameters).deliver!
+            when 'new'     then nil
+            else super
+          end
+        else
+          super
         end
       end
 
@@ -440,7 +443,7 @@ module ActionMailer #:nodoc:
       private
         def matches_dynamic_method?(method_name) #:nodoc:
           method_name = method_name.to_s
-          /(create|deliver)_([_a-z]\w*)/.match(method_name) || /^(new)$/.match(method_name)
+          /^(create|deliver)_([_a-z]\w*)/.match(method_name) || /^(new)$/.match(method_name)
         end
     end
 
diff --git a/actionmailer/test/mail_service_test.rb b/actionmailer/test/mail_service_test.rb
index 7f9540c..c16444c 100644
--- a/actionmailer/test/mail_service_test.rb
+++ b/actionmailer/test/mail_service_test.rb
@@ -1031,4 +1031,8 @@ class RespondToTest < Test::Unit::TestCase
   def test_should_not_respond_to_deliver_with_template_suffix_if_it_begins_with_a_digit
     assert !RespondToMailer.respond_to?(:deliver_1_template)
   end
+
+  def test_should_not_respond_to_method_where_deliver_is_not_a_suffix
+    assert !RespondToMailer.respond_to?(:foo_deliver_template)
+  end
 end
-- 
1.6.0.2

